### ISSUE TITLE
Any chance of python 3 conda package?

### ISSUE THREAD

**Issue by mj-harvey** [psi4/psi4#466]
Hi,

I see you have a 1.0.54 psi4 package in conda. THanks! 
would it be possible to have a package for Python 3.5 as well? (Asking out of ignorance, I don't know what the current state of PSI4's  Python 3 compatibility is)

Matt

**Comment by loriab** [comment]
At various times Psi4 has been made python3-compatible, and then that compatibility diminishes as the developers don't use py3. We're in the middle of an infrastucture update at present (build and python bindings). After that, python3 support and/or switch is likely, so probably this calendar year. I'm sorry it probably can't be this month, but there's so many under-the-hood changes that we want to decouple them as much as possible. A py35 package is definitely a target (and will be far easier when we don't have to deal with Boost).

**Comment by mj-harvey** [comment]
Ok, thanks for the info. I'll see about building one myself, and submit a PR with any changes required for Python 3.

**Comment by CDSherrill** [comment]
Due to the very large number of changes that are to be pushed in the near
future, you might want to hold off working on this for a few days.

**Comment by loriab** [comment]
I strongly concur. If you do anything now, you'll have to deal with Boost::python, whereas the current development head (https://github.com/jturney/psi4/commits/CMakeRewrite-KillTheBoost) uses pybind11 for python binding. I think the new plan is to merge that branch (which is 250 commits ahead) back into master this week with main functionality intact but some lesser items subject to polish and change (e.g., docs, cmake, plugins, psimrcc-fd-freq2).

**Comment by loriab** [comment]
Ok, the big merge is in. If you'd like to work on Py3/conda, that'd be great. Note that the python bindings are now through pybind11 and there is (this issue)[https://github.com/pybind/pybind11/pull/227] that I haven't looked into. The existing conda recipes for  psi4 are (here)[https://github.com/psi4/psi4meta/tree/master/conda-recipes] and though they were in pristine shape at the beginning of July, they haven't had CMakeRewrite/KillTheBoost applied to them yet. And ignore the `LIBC_INTERJECT` business.

Basically, if you'd like to work on this, wonderful, and if you have any problems, ping me. But in any case, I will get to this eventually (see, #111).

**Comment by loriab** [comment]
Except for some stray bits (sow/reap, databases, the ancillary scripts directory), Psi4 is now python3 compatible (largely thanks to @dgasmith). In fact, the majority of our CI testing is now with py3, so it should stay that way.

There is now a py3 conda package for Linux uploaded to anaconda.org. Niceties of packaging and channels aren't in place yet, but a full working psi4/py3 conda environment (named `p4py35`) can be obtained with the below. Please report any problems.

`conda create -n p4py35 python=3.5 psi4 numpy dftd3 -c psi4/label/test -c psi4`

**Comment by loriab** [comment]
Mac py3 package via the below. Note that it's clang, not gcc (until I find a working set of conda mac gcc >=4.9 compilers), so no parallelism.

`conda create -n p4py35 python=3.5 psi4 numpy dftd3 -c psi4/label/clang -c psi4`

**Comment by loriab** [comment]
FYI, new set of conda packages today. These handle the Linux "hello world" bug internally. It's recommended _not_ to update from a circa 1.0 psi4 (e.g., 1.0.54) with `conda update psi4` but to create a new environment for them as follows. If you already have a pre 1.1 psi4 installed, `conda update psi4 -c psi4/label/test -c psi4` should do the trick.

* Linux, py27: `conda create -n p4env python=2.7 psi4 numpy dftd3 -c psi4/label/test -c psi4`
* Linux, py35: `conda create -n p4env python=3.5 psi4 numpy dftd3 -c psi4/label/test -c psi4`
* Mac, py27: `conda create -n p4env python=2.7 psi4 numpy dftd3 -c psi4/label/clang -c psi4`
* Mac, py35: `conda create -n p4env python=3.5 psi4 numpy dftd3 -c psi4/label/clang -c psi4`

**Comment by j3mdamas** [comment]
Hi guys,
What is the status on this? I see [here](https://anaconda.org/psi4/psi4/files) that the dev has the py35 and py36. Any idea of when this dev version will be definitive (released)?

**Comment by loriab** [comment]
Yes, there's now py27, py35, py36 for Linux and Mac. I've been getting the build and packaging automated again these past few days. There's a few more external projects that need final integration (simint, ambit, a slight dkh fix). I'm aiming for RC1 by mid next week because I'll be away for a bit after that.

**Comment by j3mdamas** [comment]
Thanks @loriab. We're fine with RC, I just didn't want to start testing the dev version.

**Comment by loriab** [comment]
A few docs

What's in the conda packages psi4 & psi4-deps and the installer psi4conda?
--------------------------------------------------------------------------

psi4 - has full-featured psi4 itself and necessarily all the link-time qc addons (e.g., chemps2). Of gcc-ness, it has minimal, run-time libraries, not compilers, though, for linux, full gcc and run-time gcc are the same. It doesn't have the run-time qc addons (e.g., dftd3).

psi4-deps - does not have psi4 itself (though fine to install it alongside). Does have all the link-time and run-time addons. Does have cmake and sphinx (and python). Of gcc-ness, has full packages, that is, compilers, not runtime packages.

installer - has full-featured psi4 itself, all link-time qc addons, all run-time qc addons, and minimal gcc runtime libraries.


How to just get Psi4 with minimal conda interaction
---------------------------------------------------

Download one of the six installers (Linux/Mac; Py27, 35, 36; presently at http://vergil.chemistry.gatech.edu/download/). `bash` it. Follow the prompts and _do_ make the adjustments to PATH and PSI_SCRATCH that it suggests at the end. Test with `psi4 --test`. Done.


How to get a pre-build Psi4 into a conda environment of an existing Ana/Miniconda installation
----------------------------------------------------------------------------------------------

`conda create -n p4env python=x.x psi4 psi4-deps -c psi4/label/devel -c psi4`

Same for Mac/Linux. Substitute desired python version: 2.7, 3.5, 3.6. Only reason for psi4-deps is to get the qc runtime addons; could say "dftd3 gcp v2rdm_casscf" instead of "psi4-deps"; or leave them out if you don't want them. Activate environment and make the adjustmets to PATH and PSI_SCRATCH that it suggests at the end. Test with `psi4 --test`.


How to use all those conda packages to compile Psi4 faster and easier
---------------------------------------------------------------------

`conda create -n p4deps python=x.x psi4-deps -c psi4`

Same for Mac/Linux. Substitute desired python version: 2.7, 3.5, 3.6. Fine to choose your own env name. Activate environment, `source activate p4deps`. Go to where you've cloned psi4. Execute `psi4-path-advisor`. It gives you a basic cmake command covering python, sphinx, link-time qc addons, and run-time qc addons. There's a help menu `-h` that gives more info. There's other options that will also pre-configure compilers. For example, at GaTech `psi4-path-advisor --intel` works. On Macs with XCode, `psi4-path-advisor --clang` works. Just read the help. For DGAS, there's an "--disable"", but I don't encourage it. It gives you a fully functional cmake command, but those are just setting up CMake cache - like the plugins you can always add your own CMake variables to the command.


Plugins from a conda Psi4
-------------------------

Works on Linux. I need to adjust at least cache_pybind11_DIR on Mac.






How to bump a version
---------------------

* 0. **ACTION** Check everything in.

* 1. **OBSERVE** Current versioning state.

  * Be on master of (i) a direct clone or (ii) clone-of-fork with master up-to-date with upstream (including tags!!!) and with upstream as remote.

  * https://github.com/psi4/psi4/releases says v1.1a1 007a9b6

  ```
>>> git tag
v1.0
v1.1a1

>>> cat psi4/metadata.py 
__version__ = '1.1a1'
__version_long = '1.1a1+007a9b6'
__version_upcoming_annotated_v_tag = '1.1a2'

>>> git describe --abbrev=7 --long --always HEAD
v1.1a1-417-gcbee32b

>>> git describe --abbrev=7 --long --dirty
v1.1a1-417-gcbee32b

>>> ./psi4/versioner.py 
Defining development snapshot version: 1.1a2.dev417+cbee32b (computed)
1.1a2.dev417 {master} cbee32b 1.0.0.999   1.0 <-- 1.1a2.dev417+cbee32b

>>> git diff

  ```

  * Observe that current latest tag matches metadata scipt and git describe, that GH releases matches metadata script, that upcoming in metadata script matches current versioner version.

  * Note that current tag is "v1.1a1". Decide on imminent tag, say "v1.1rc1".

* 2. **ACTION** Bump tag in code.

  * Edit current & prospective tag in psi4/metadata.py. Use your decided-upon tag "v1.1rc1" and a speculative next tag, say "v1.1rc2", and use 7 "z"s for the part you can't predict.

  ```
>>> vi psi4/metadata.py

>>> git diff
diff --git a/psi4/metadata.py b/psi4/metadata.py
index 5d87b55..6cbc05e 100644
--- a/psi4/metadata.py
+++ b/psi4/metadata.py
@@ -1,6 +1,6 @@
-__version__ = '1.1a1'
-__version_long = '1.1a1+007a9b6'
-__version_upcoming_annotated_v_tag = '1.1a2'
+__version__ = '1.1rc1'
+__version_long = '1.1rc1+zzzzzzz'
+__version_upcoming_annotated_v_tag = '1.1rc2'

>>> git add psi4/metadata.py

>>> git commit -m "v1.1rc1"
  ```

* 3. **OBSERVE** Undefined version state.

  ```
>>> git describe --abbrev=7 --long --always HEAD
v1.1a1-418-g6100822

>>>  git describe --abbrev=7 --long --dirty
v1.1a1-418-g6100822

>>>  psi4/versioner.py 
Undefining version for irreconcilable tags: 1.1a1 (computed) vs 1.1rc1 (recorded)
undefined {master} 6100822 1.0.0.999   1.0 <-- undefined+6100822
  ```

  * Note 7-char git hash for the new commit, here "6100822".

* 4. **ACTION** Bump tag in git. Bump git tag in code.

  * Use the decided-upon tag "v1.1rc1" and the observed hash "6100822" to mint a new *annotated* tag, minding that "v"s are present here.

  * Use the observed hash to edit psi4/metadata.py and commit immediately.

  ```
>>> git tag -a v1.1rc1 6100822 -m "v1.1rc1"

>>> vi psi4/metadata.py 
>>> git diff
diff --git a/psi4/metadata.py b/psi4/metadata.py
index 6cbc05e..fdc202e 100644
--- a/psi4/metadata.py
+++ b/psi4/metadata.py
@@ -1,5 +1,5 @@
 __version__ = '1.1rc1'
-__version_long = '1.1rc1+zzzzzzz'
+__version_long = '1.1rc1+6100822'
 __version_upcoming_annotated_v_tag = '1.1rc2'

>>> psi4/versioner.py 
Amazing, this can't actually happen that git hash stored at git commit.
>>> git add psi4/metadata.py
>>> git commit -m "Records tag for v1.1rc1"
  ```

* 5. **OBSERVE** Current versioning state.

  * Nothing to make note of, this is just a snapshot.

  ```
>>> psi4/versioner.py 
Defining development snapshot version: 1.1rc2.dev1+4e0596e (computed)
1.1rc2.dev1 {master} 4e0596e 1.0.0.999   1.0 <-- 1.1rc2.dev1+4e0596e

>>> git describe --abbrev=7 --long --always HEAD
v1.1rc1-1-g4e0596e

>>> git describe --abbrev=7 --long --dirty
v1.1rc1-1-g4e0596e

>>> git tag
v1.0
v1.1a1
v1.1rc1

>>> cat psi4/metadata.py 
__version__ = '1.1rc1'
__version_long = '1.1rc1+6100822'
__version_upcoming_annotated_v_tag = '1.1rc2'

>>> cat metadata.out.py | head -8
__version__ = '1.1rc2.dev1'
__version_branch_name = 'master'
__version_cmake = '1.0.0.999'
__version_is_clean = 'True'
__version_last_release = '1.0'
__version_long = '1.1rc2.dev1+4e0596e'
__version_prerelease = 'False'
__version_release = 'False'

>>> git log --oneline
4e0596e Records tag for v1.1rc1
6100822 v1.1rc1
cbee32b Fixes pcmsolver/scf for py3. Moves source for libefp upstream.
  ```

6. **ACTION** Move bump to remote.

  * Temporarily disengage "Include administrators" on protected master branch.

  ```
>>> git push origin master
>>> git push origin v1.1rc1
  ```

  * Now https://github.com/loriab/psi4/releases says v1.1rc1 6100822

Notes
-----

* To delete a tag:

  * >>> git tag -d v1.1a1
  * >>> git push origin :refs/tags/v1.1a1

**Comment by loriab** [comment]
Python 3 conda packages have been around for a while now. All the docs there are transitioned into the permanent docs. Nightly builds can be found (for both OS) on `-c psi4/label/dev`.

**Comment by j3mdamas** [comment]
Hi @loriab,

When will it change from dev to the normal channel?

**Comment by loriab** [comment]
Ah, the plan is to have only official releases through `conda install psi4 -c psi4`. Any nightly build will be through `conda install psi4 -c psi4/label/dev [-c psi4]` (bracketed part needed if any QC dependencies need to be got from the main psi4 channel). This is in the docs, but not any that have built yet :-)

So next week there'll be a 1.1 on main psi4 channel. We're gearing up for `1.1rc2` now.

**Comment by j3mdamas** [comment]
Right, I saw the publication on Ahead of print on JCTC, so I guess things are going well. Great news, thanks!

**Comment by loriab** [comment]
Thanks, good to know other projects are interested.
